c++/src/kj/compat/http.h

```diff
@@ -662,9 +659,20 @@ class HttpClient {
   // `url` and `headers` need only remain valid until `openWebSocket()` returns (they can be
   // stack-allocated).
 
-  virtual kj::Promise<kj::Own<kj::AsyncIoStream>> connect(kj::StringPtr host);
-  // Handles CONNECT requests. Only relevant for proxy clients. Default implementation throws
-  // UNIMPLEMENTED.
+  struct ConnectResponse {
+    uint statusCode;
+    kj::StringPtr statusText;
+    const HttpHeaders* headers;
+    kj::Own<kj::AsyncIoStream> io;
```

Comment:
If the server response with an error, is the connection still established? I'd expect not. Can the error response contain a regular HTTP response body? If so then I think we need a `OneOf` here to allow for that, like in `WebSocketResponse`.

Reply:
If the server responds with an error, no the connection is not established, and yes, the error response can contain a regular HTTP response body.

Specifically from the spec:

```
   Any 2xx (Successful) response indicates that the sender (and all
   inbound proxies) will switch to tunnel mode immediately after the
   blank line that concludes the successful response's header section;
   data received after that blank line is from the server identified by
   the request-target.  Any response other than a successful response
   indicates that the tunnel has not yet been formed and that the
   connection remains governed by HTTP.
```

Which is fun given that a 204 No Content, etc will still indicate that there *will* be an ongoing payload despite the normal rules for those responses. If an error response is returned, then the connection is still an HTTP connection and HTTP rules still apply.

Reply:
OK, so we do indeed need this to be:

`kj::OneOf<kj::Own<kj::AsyncIoStream>, kj::Own<kj::AsyncInputStream>> connectionOrErrorBody`

The latter would be used in the case of an error response.